pkg/adsc/delta.go

```diff
@@ -0,0 +1,739 @@
+func Register[T proto.Message](f func(ctx HandlerContext, res T, event Event)) Option {
+	return func(c *Client) {
+		c.handlers[typeName[T]()] = func(ctx HandlerContext, res proto.Message, event Event) {
+			f(ctx, res.(T), event)
+		}
+	}
+}
+
+// Watch registers an initial watch for a type based on the type reflected by the proto message.
+func Watch[T proto.Message](resourceName string) Option {
+	return initWatch(typeName[T](), resourceName)
+}
+
+func initWatch(typeURL string, resourceName string) Option {
+	return func(c *Client) {
+		if resourceName == "*" {
+			// Normalize to allow both forms
+			resourceName = ""
+		}
+		key := resourceKey{
+			Name:    resourceName,
+			TypeURL: typeURL,
+		}
+		existing, f := c.tree[key]
+		if f {
+			// We are watching directly now, so erase any parents
+			existing.Parents = nil
+		} else {
+			c.tree[key] = resourceNode{
+				Parents:  make(keySet),
+				Children: make(keySet),
+			}
+		}
+		c.initialWatches = append(c.initialWatches, key)
+	}
+}
+
+func (c *Client) handleRecv() {
+	// We connected, so reset the backoff
+	if c.cfg.BackoffPolicy != nil {
+		c.cfg.BackoffPolicy.Reset()
+	}
+	for {
+		deltaLog.Infof("Start Recv for node %v", c.nodeID)
+		msg, err := c.xdsClient.Recv()
+		if err != nil {
+			deltaLog.Infof("Connection closed for node %v with err: %v", c.nodeID, err)
+			select {
+			case c.errChan <- err:
+			default:
+			}
+			// if 'reconnect' enabled - schedule a new Run
+			if c.cfg.BackoffPolicy != nil {
+				time.AfterFunc(c.cfg.BackoffPolicy.NextBackOff(), c.reconnect)
+			} else {
+				c.Close()
+				c.WaitClear()
+				c.deltaXDSUpdates <- nil
+				close(c.errChan)
+			}
+			return
+		}
+		deltaLog.Infof("Received response: %s", msg.TypeUrl)
+		if err := c.handleDeltaResponse(msg); err != nil {
+			deltaLog.Infof("Handle response %s failed: %v", msg.TypeUrl, err)
+			c.Close()
+			return
+		}
+		c.mutex.Lock()
+		c.received[msg.TypeUrl] = msg
+		c.mutex.Unlock()
+		c.deltaXDSUpdates <- msg
+	}
+}
+
+func (c *Client) handleDeltaResponse(d *discovery.DeltaDiscoveryResponse) error {
+	var rejects []error
+	allAdds := map[string]set.Set[string]{}
+	allRemoves := map[string]set.Set[string]{}
+	ctx := &handlerContext{}
+	if isDebugType(d.TypeUrl) {
+		// No need to ack and type check for debug types
+		return nil
+	}
+	for _, r := range d.Resources {
+		if d.TypeUrl != r.Resource.TypeUrl {
+			deltaLog.Errorf("Invalid response: mismatch of type url: %v vs %v", d.TypeUrl, r.Resource.TypeUrl)
+			continue
+		}
+		err := c.trigger(ctx, d.TypeUrl, r, EventAdd)
+		if err != nil {
+			return err
+		}
+		parentKey := resourceKey{
+			Name:    r.Name,
+			TypeURL: r.Resource.TypeUrl,
+		}
+		c.establishResource(parentKey)
+		if ctx.nack != nil {
+			rejects = append(rejects, ctx.nack)
+			// On NACK, do not apply resource changes
+			continue
+		}
+
+		remove, add := c.tree[parentKey].Children.Diff(ctx.sub)
```

Comment:
Confused about the `remove`.  DeltaResponse has a removed field, why do you need to compare local cache with the response to get the removed resources

Reply:
NVM, got it.